yarn-project/archiver/src/archiver/archiver.ts

```diff
@@ -1246,6 +1327,202 @@ export class Archiver extends (EventEmitter as new () => ArchiverEmitter) implem
+    }
+
+    // For other blocks, check if any subsequent block references our block as parent
+    const ourLocalHeaderHash = localPendingBlock.header.toPropose().hash().toString();
+
+    // Look for the next block that should reference our block as parent
+    const nextBlockEvent = blockEvents.get(localPendingBlockNumber + 1);
+    if (nextBlockEvent) {
+      const nextBlockParentHash = nextBlockEvent.args.parentHeaderHash;
+
+      if (nextBlockParentHash !== ourLocalHeaderHash) {
+        this.log.warn(`Parent chain discontinuity detected`, {
+          localPendingBlockNumber,
+          ourLocalHeaderHash,
+          nextBlockNumber: localPendingBlockNumber + 1,
+          nextBlockParentHash,
+        });
+        return false;
+      }
+    }
+
+    // Also check if our block's expected parent exists and matches
+    if (localPendingBlockNumber > 1) {
+      const parentBlock = await this.getBlock(localPendingBlockNumber - 1);
+      const parentEvent = blockEvents.get(localPendingBlockNumber - 1);
+
+      if (parentBlock && parentEvent) {
+        const expectedParentHash = parentBlock.header.toPropose().hash().toString();
+        const ourBlockEvent = blockEvents.get(localPendingBlockNumber);
+
+        if (ourBlockEvent && ourBlockEvent.args.parentHeaderHash !== expectedParentHash) {
+          this.log.warn(`Our block's parent header hash doesn't match expected parent`, {
+            localPendingBlockNumber,
+            expectedParentHash,
+            actualParentHash: ourBlockEvent.args.parentHeaderHash,
+          });
+          return false;
+        }
+      }
+    }
+
+    return true; // Chain continuity verified
+  }
+
+  /**
+   * Handles retroactive proof updates for blocks that were synced before they were proven.
+   * When blocks are initially synced without proofs, they get stored with archive.root = Fr.ZERO.
+   * Later when they get proven, we need to update the stored blocks with the real archive roots.
+   */
+  private async handleRetroactiveProofUpdates(currentL1BlockNumber: bigint): Promise<void> {
+    try {
+      // Start searching from the last proven block's L1 location, since we want to find
+      // proofs for blocks that were synced but not yet proven
+      const lastProvenL2BlockNumber = await this.store.getProvenL2BlockNumber();
+      let searchFromL1Block: bigint;
+
+      if (lastProvenL2BlockNumber > 0) {
+        // Get the L1 block number where the last proven L2 block was published
+        const lastProvenBlock = await this.store.getPublishedBlock(lastProvenL2BlockNumber);
+        searchFromL1Block = lastProvenBlock?.l1.blockNumber ?? this.l1constants.l1StartBlock;
+      } else {
+        // No blocks proven yet, start from L1 genesis
+        searchFromL1Block = this.l1constants.l1StartBlock;
+      }
+
+      // Get all proof events from the L1 range starting after the last proven block
+      const provenBlocks = await retrieveL2ProofsFromRollup(
```

Comment:
Would you not sometimes end up looking over the same range as you have already fetched events for, but because of the lack of overlap only find them here and not earlier?